### PR TITLE
Port FFTs to gpuarray backend

### PR DESCRIPTION
**Pull request by slefrancois**

### CODE REVIEW

File: theano/gpuarray/fft.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+
+from theano.gpuarray import (basic_ops, GpuArrayType)
+
+try:
+    import pygpu
+    pygpu_available = True
+except ImportError:
+    pygpu_available = False
+
+import pycuda.driver
```
Comment (abergeron) [inline theano/gpuarray/fft.py]: This import has to be guarded.

**Comment by abergeron** [comment]
Ok, modulo my comment.

**Comment by slefrancois** [comment]
Done, it was also producing a Travis error.

**Comment by slefrancois** [comment]
I pushed preliminary gradient code that wasn't ready, new commit fixes that.

File: theano/gpuarray/fft.py [hunk 2]
```diff
@@ -121,33 +129,23 @@ def thunk():
 
         return thunk
 
-cufft = CuFFTOp()
```
Comment (nouiz) [inline theano/gpuarray/fft.py]: do:
curfft = CuRFFTOp()

and reuse that in the method in this file.

File: theano/gpuarray/fft.py [hunk 3]
```diff
@@ -230,13 +235,34 @@ def thunk():
 
         return thunk
 
-cuifft = CuIFFTOp()
```
Comment (nouiz) [inline theano/gpuarray/fft.py]: idem, make curifft here

**Comment by slefrancois** [comment]
Hold off on merging, there are issues with normalizations and real FFTs when it comes to gradients. We might have to change the interface.

**Comment by slefrancois** [comment]
Ready to merge.

File: doc/library/gpuarray/fft.txt [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+==============================================
+
+Performs Fast Fourier Transforms (FFT) on the GPU.
+
+FFT gradients are implemented as the opposite Fourier transform of the output gradients.
+
+.. warning ::
+    The real and imaginary parts of the Fourier domain arrays are stored as a pair of float32
+    array, emulating complex64. Since theano does not support complex
+    number operations, care must be taken to manually implement operators such as complex
```
Comment (nouiz) [inline doc/library/gpuarray/fft.txt]: Theano support them in limited way. But we don't support the grad on it. So maybe tell that we don't support grad on complex?

File: theano/gpuarray/fft.py [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+                                       batch=input_shape[0])
+
+                # Sync GPU variables before computation
+                input_pycuda.sync()
+                output_pycuda.sync()
+
+                fft.fft(input_pycuda, output_pycuda, plan[0])
+
+                # Sync results to ensure output contains completed computation
+                pycuda.driver.Context.synchronize()
```
Comment (nouiz) [inline theano/gpuarray/fft.py]: @abergeron is this needed? Can we add an event in the output variable? I would let this for the futur, but I would like to know. We probably need to make an issue about which op aren't async.

**Comment by nouiz** [comment]
I just did a quick overview of this PR.

**Comment by slefrancois** [comment]
New commit to skip test requiring GPU libraries.

**Comment by slefrancois** [comment]
This is ready to review and merge.

It would be worth checking if the gpu op_lifters are as we want. I did it so that they are ignored if scikits_cuda is not present. However, I have seen other gpu ops that register the optimization, but return an error in the lifter if optional dependencies are absent.

**Comment by nouiz** [comment]
travis have some failure.

File: theano/gpuarray/fft.py (hunk 5 above)
Comment (abergeron) [inline theano/gpuarray/fft.py]: This might not be required anymore, now that no stream in libgpuarray is async with regards to the null stream.

**Pull request merged**